### ISSUE TITLE
LBANN help out of date

### ISSUE THREAD

**Issue by benson31** [LLNL/lbann#864]
It has been noted in various fora lately that `lbann --help` is out-of-date. Those in the know of the actual options that LBANN supports should update this output.

**Comment by benson31** [comment]
Yes, I'm spamming everyone. Sorry.

**Comment by ndryden** [comment]
Our options parsing package should be able to automatically generate basic information for `--help` based on the options it parses. If it can't, we should move on from whatever options package we're using.

**Comment by benson31** [comment]
I'm refering to the message that's manually written and printed [here](https://github.com/LLNL/lbann/blob/4a0d1bedfb8212d299f1a1a583391342c799dd74/src/proto/proto_common.cpp#L797). There is nothing automatic about it. We don't use an external package (we probably should; I like [clara](https://github.com/catchorg/clara)). The `options` class is declared here in `include/lbann/utils/options.hpp`.

**Comment by ndryden** [comment]
Yeah. That should not exist, and our options class should self-document. We can add a short blurb at the top.

**Comment by davidHysom** [comment]
The options class permits you to add/delete items from anyplace inside lbann. You can also ask, "do have in integer for the key "foo?" If yes please return it, if not here is the default value to use for the remainder of this executable's lifetime.


It also parses the command line, and can optionally parse keys/values that are stored in a user-specified file.


I'm unsure how --help would work for self-documentation, since "options" may not know all keys/values at runtime startup. And if you want it to


I find our current options class useful for developing/experimenting with new algorithms. If I'm developing something that -- if it works and we decide to keep it -- should be added to our prototext. However, I don't want to continually be changing that, so instead I'll query  "options" for values.


And if you want it to self-document, does that mean we'd need to pass descriptions on the command line, e.g:

  --data_store_mode=0 {Determines the algorithm that data_stores use to exchange data. Legal values are 0: blah, blah; 1: blah, blah}


If people are unhappy enough to want to change how "options" works, or to use a different package, I suggest this needs a lot more thought/agreement as to what we really need.

________________________________

**Comment by benson31** [comment]
The list of options should be gathered dynamically, based on the options exposed by the exe in question. That is, `lbann` probably has different options than `lbann_inf`, etc.

Code for this generally looks like:
```cpp
int main(int argc, char* argv[])
{
    ArgumentParser args;
    args.AddOption("-dc", "--disable-cuda", "Disable CUDA support in LBANN"); // Flag
    args.AddOption("-nt","--num-io-threads",1UL,"Number of threads for I/O"); // Value
    // More options

    try {
        args.parse(argc, argv);
    }
    catch (PrintHelpException const&)
    {
        if (rank==0)
            args.PrintHelp(std::err);
        return EXIT_SUCCESS;
    }

    auto&& num_io_threads = args.template Get<size_t>("num-io-threads");
    bool disable_cuda = args.template Get<bool>("disable-cuda");
    // Rest of main()
}
```
For this example, `AddOption` is some overloaded function that distinguishes flag (boolean interpreted as `true` if there, `false` if not) from options that have value:

```cpp
void ArgumentParser::AddOption(
    std::string short_flag, std::string long_flag, std::string doc_string);
template <typename T>
void ArgumentParser::AddOption(
    std::string short_flag, std::string long_flag, T&& default_value,
    std::string doc_string);
```
Note that this gives static types to all arguments. By using a type-safe type-erased map (`std::unordered_map<std::string, std_cxx17::any>`; `std::any` is C++17, `any` was implementable in C++98), type safety can be maintained on the user side. One could either expose another argument for the key into this map or it can just be defined as the `long_flag` trimmed of any leading dashes, for example. Moreover, types are checked during parsing so if you do something like `lbann --num-io-threads potato`, you'll get a nice exception telling you that parsing failed becaues "potato" could not be converted to `size_t` (or {`unsigned`, `unsigned short`, `unsigned char`} if you're _really_ sure you'll never have more than {4,294,967,296; 65536; 256} I/O threads/rank :P).

If there are arguments that are truly universal, such as `-h/--help` (`--disable-cuda` is probably universal w.r.t. LBANN but I wanted a flag that I knew to illustrate adding a flag...), they can be added by default to the argument parser in its constructor. When `PrintHelp()` is called, the internal structures loop over known options, pretty-printing short-flag, long-flag, default value (if one exists), and the application-defined doc string.

This would give flexibility to the parsing framework to allow multiple executables with unique needs to represent their options independently and print helpful messages.

**Comment by davidHysom** [comment]
Hm ... so can I ad-hock add an option on the cmd line

for which there is no "args.AddOption" in main?

And/or, can I sprinkle "args.AddOption" anyplace I want -- i.e, does it have to be in main?

________________________________

**Comment by benson31** [comment]
This would be implemented as a policy so it's easy to change behavior (or even behaviour as my autocorrect is insisting). The issue is strong typing of arguments (fairly common in argument parsers (boost and clara are the two with which I'm familiar, and both have strong typing)). Unknown arguments must then either throw some `UnknownOptionError` or something, be stored as strings (which probably requires all arguments to take values, which is...not great), or be ignored. Of these, the first or the last options are my general preference; however, I would implement this theoretical ArgumentParser with a policy parameter to allow any of these options.

Before continuing, I would strongly advocate putting all arguments in `main()` if for no other reason than they'd all be in one place and easily found -- users don't complain about not finding a particular option, etc. Also, it ensures that `PrintHelp()` is complete and accurate (that will only be called once and whatever is registered by then is what gets printed).

That said, we can do whatever we want/need. If we ignore extra arguments, one could call `AddOption()` at any point. You'd just have to re-parse. Or the parser could operate slightly differently, namely one could construct it with `argc, argv`, which `AddOption` would search and populate the map as it goes. This does not solve the `PrintHelp()` problem, but if these are temporary options that you're using in some development branch that will migrate to the appropriate `main()`s when the feature goes public, this could be a viable approach. (Though, again, I might argue that if you can edit one file, you can probably edit a second file and make a temporary change either to the exe you're using or to the argument parser's global/default arguments...)

**Comment by benson31** [comment]
Also, while this is probably a conversation we should have about improving some of the infrastructural things, I don't really care right now. Updating the help message that's currently printed is sufficient. I just don't want users complaining that they tried something they found in `lbann --help` and it either did nothing or caused unexpected behavior. That's the real problem I'm trying to address in this issue.

**Comment by davidHysom** [comment]
Tom, I agree 100%.

________________________________

**Comment by davidHysom** [comment]
I would like to modify the existing options class as follows. Currently the help message is hard-coded in c++. I would like it to come from the options class. Actually, the options class will read a file that contains keys, values, types, and a description for known options. When construction the model/reader/optimizer the code in src/proto can query the options (I'm leaving out  details  as to how this works. We retain the extremely useful ability to add and rertrieve options on-the-fly, Modifying the text file should considerably ease the burden of keeping the help message up to date.

**I cannot express how strongly I am opposed to putting all options in main** I used to work with Trilinos in a different project, and that's what they do. One of my primary motivations for the options class was to get away from that nonsense.
OK, if you have a mature, reasonably stable code it might be a reasonable approach. But we're changing things all the time -- e.g, adding, modifying, and deleting layer types. And many of those have settings that users should have the ability to modify on the cmd line (as opposed to modifying the prototext files). We should not have to remember to update the various main() when we add or delete layer types.

If no one objects to my changes proposed above, I'll start a PR.

**Comment by bvanessen** [comment]
I would like to talk about if we should continue to roll our own here or to use an outside tool like https://github.com/catchorg/clara

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470
